src/reporter.cc

```diff
@@ -71,46 +78,109 @@ void MetadataReporter::ReportMetadata() {
+    // us to sent a request to the batch endpoint when we have a single request.
+    const json::Object* single_request = entries[0]->As<json::Object>();
+    json::value duplicate_request = json::object({  // ResourceMetadata
+      {"name", json::string(single_request->Get<json::String>("name"))},
+      {"type", json::string(single_request->Get<json::String>("type"))},
+      {"location", json::string(single_request->Get<json::String>("location"))},
+      {"state", json::string(single_request->Get<json::String>("state"))},
+      {"eventTime",
+        json::string(single_request->Get<json::String>("eventTime"))},
+    });
+    entries.emplace_back(std::move(duplicate_request));
   }
-
+  const std::string content_type =
+      std::string("multipart/mixed; boundary=") + kMultipartBoundary;
   http::client client;
-  http::client::request request(endpoint);
-  std::string request_body = update_metadata_request->ToString();
+  http::client::request request(batch_uri);
   request << boost::network::header("User-Agent", user_agent);
-  request << boost::network::header("Content-Length",
-                                    std::to_string(request_body.size()));
-  request << boost::network::header("Content-Type", "application/json");
+  request << boost::network::header("Content-Type", content_type);
   request << boost::network::header("Authorization", auth_header);
-  request << boost::network::body(request_body);
+  request << boost::network::header("Expect", "100-continue");
+  std::ostringstream out;
+  out << std::endl;
+
+  for (json::value& entry : entries) {
+    const json::Object* single_request = entry->As<json::Object>();
+    std::string request_body = single_request->ToString();
+    out << "--" << kMultipartBoundary << std::endl;
+    out << "Content-Type: application/http" << std::endl;
+    out << "Content-Transfer-Encoding: binary" << std::endl;
+    out << "Content-ID: " << single_request->Get<json::String>("name")
+        << std::endl;
+    out << std::endl;
+    out << "POST " << publish_path << std::endl;
+    out << "Content-Type: application/json; charset=UTF-8" << std::endl;
+    out << "Content-Length: " << std::to_string(request_body.size())
+        << std::endl;
+    out << std::endl << request_body << std::endl;
+  }
+  out << "--" << kMultipartBoundary << "--" << std::endl;
+  std::string multipart_body = out.str();
+  const int total_length = multipart_body.size();
+
+  request << boost::network::header("Content-Length",
+                                    std::to_string(total_length));
+  request << boost::network::body(multipart_body);
+
+  if (verbose_logging) {
+    LOG(INFO) << "About to send request: POST " << batch_uri;
+    LOG(INFO) << "Headers:";
+    http::client::request::headers_container_type head = (headers(request));
+    for (auto it = head.begin(); it != head.end(); ++it) {
+      if (it->first == "Authorization") {
```

Comment:
Why are we excluding the token? It's being logged in other places as well, so I wouldn't bother. Just dump all the headers, i.e.:
```c++
    LOG(INFO) << "Headers: " << headers(request);
```

Reply:
Actually, it's `request.headers()`, see, e.g., [oauth2.cc](https://github.com/Stackdriver/metadata-agent/blob/master/src/oauth2.cc#L230)... You'll also need to `#include "http_common.h"`.

Reply:
done